### PR TITLE
Eligibility result now stored in session

### PR DESCRIPTION
**Pull request by AndyHWChung**

### CODE REVIEW

File: app/uk/gov/hmrc/helptosavefrontend/controllers/NewApplicantUpdateEmailAddressController.scala [hunk 1]
```diff
@@ -33,8 +32,10 @@ import uk.gov.hmrc.helptosavefrontend.config.{FrontendAppConfig, FrontendAuthCon
 import uk.gov.hmrc.helptosavefrontend.connectors.{EmailVerificationConnector, SessionCacheConnector}
 import uk.gov.hmrc.helptosavefrontend.forms.EmailVerificationErrorContinueForm
 import uk.gov.hmrc.helptosavefrontend.metrics.Metrics
+import uk.gov.hmrc.helptosavefrontend.models.HTSSession.EligibleWithUserInfo
 import uk.gov.hmrc.helptosavefrontend.models._
-import uk.gov.hmrc.helptosavefrontend.models.eligibility.IneligibilityType
+import uk.gov.hmrc.helptosavefrontend.models.eligibility.EligibilityCheckResult.{AlreadyHasAccount, Eligible, Ineligible}
+import uk.gov.hmrc.helptosavefrontend.models.eligibility.IneligibilityReason
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/NewApplicantUpdateEmailAddressController.scala]: unused import ?

File: app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala [hunk 2]
```diff
@@ -161,7 +165,13 @@ class RegisterController @Inject() (val messagesApi:           MessagesApi,
       checkIfDoneEligibilityChecks { eligibleWithEmail ⇒
         eligibleWithEmail.confirmedEmail.fold[Future[Result]](
           SeeOther(routes.RegisterController.getSelectEmailPage().url))(
-            _ ⇒ Ok(views.html.register.create_account_help_to_save()))
+            _ ⇒
+              EligibilityReason.fromEligible(eligibleWithEmail.eligible).fold{
+                logger.warn(s"Could not parse eligiblity reason: ${eligibleWithEmail.eligible}", eligibleWithEmail.userInfo.nino)
+                internalServerError()
+              }{ reason ⇒
+                Ok(views.html.register.create_account_help_to_save(reason))
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala]: why do we need to pass the `reason` here ?

Comment (AndyHWChung) [reply app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala]: there's some content changes on this page which needs to be dynamic based on the eligibility reason - it's not being used right now but it's getting it ready to be used here

File: app/uk/gov/hmrc/helptosavefrontend/models/eligibility/EligibilityReason.scala [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+object EligibilityReason {
+
+  case object UCClaimantAndIncomeSufficient extends EligibilityReason
+
+  case class EntitledToWTC(sufficientUCIncome: Boolean) extends EligibilityReason
+
+  def fromEligible(eligible: Eligible): Option[EligibilityReason] = eligible.value.reasonCode match {
+    case 6 ⇒ Some(UCClaimantAndIncomeSufficient) // scalastyle:ignore magic.number
+    case 7 ⇒ Some(EntitledToWTC(false)) // scalastyle:ignore magic.number
```
[the remark this answers is not in the document]
Comment (AndyHWChung) [reply app/uk/gov/hmrc/helptosavefrontend/models/eligibility/EligibilityReason.scala]: just for my own reference...reason code 7 is:
```
Entitled to WTC and in receipt of positive WTC/CTC Tax Credit
```
reason code 8 is:
```
Entitled to WTC and in receipt of positive WTC/CTC Tax Credit and in receipt of UC and income sufficient
```
and the case class is:
```scala
case class EntitledToWTC(sufficientUCIncome: Boolean) extends EligibilityReason
```

[the remark this answers is not in the document]
Comment (AndyHWChung) [reply app/uk/gov/hmrc/helptosavefrontend/models/eligibility/EligibilityReason.scala]: so from the above, it looks like in reason code 7 they're either not in receipt of UC or they're in receipt of UC but their income is insufficient. If they're not in receipt of UC then technically they don't have sufficient UC I think

**Pull request merged**